### PR TITLE
[How to? pattern articles] Add record screen article

### PR DESCRIPTION
**Pull request by beaufortfrancois**
@tomayac Please have a look. Is this what you had in mind?

Live previews:

https://deploy-preview-8889--web-dev-staging.netlify.app/patterns/media/
https://deploy-preview-8889--web-dev-staging.netlify.app/patterns/media/screen-record/

### CODE REVIEW

File: src/site/content/en/patterns/media/screen-record/index.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
+---
+layout: pattern
+title: How to record screen
```
Comment (tomayac) [inline src/site/content/en/patterns/media/screen-record/index.md:3]:
```suggestion
title: How to record the user's screen
```

File: src/site/content/en/patterns/media/screen-record/index.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+date: 2022-10-14
+authors:
+  - beaufortfrancois
+description: >
+  Learn how to record screen.
```
Comment (tomayac) [inline src/site/content/en/patterns/media/screen-record/index.md:8]:
```suggestion
  Learn how to record the user's screen.
```

File: src/site/content/en/patterns/media/screen-record/index.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+height: 800
+---
+
+Sharing tabs, windows, and screens is already possible on the web platform with the [Screen Capture API](https://w3c.github.io/mediacapture-screen-share/). In short, `getDisplayMedia()` allows the user to select a screen to capture as a media stream. This stream can then be recorded with [MediaRecorder](https://developer.chrome.com/blog/mediarecorder/) or shared with others over the network. [showOpenFilePicker()](/file-system-access/) allows us finally to save the recording file locally.
```
Comment (tomayac) [inline src/site/content/en/patterns/media/screen-record/index.md:12]:
```suggestion
Sharing tabs, windows, and screens is possible on the web platform with the [Screen Capture API](https://w3c.github.io/mediacapture-screen-share/). The `getDisplayMedia()` method allows the user to select a screen to capture as a media stream. This stream can then be recorded with the [MediaRecorder API](https://developer.chrome.com/blog/mediarecorder/) or shared with others over the network. The recording can be saved to a local file via the [`showOpenFilePicker()`](/file-system-access/) method.
```

File: src/site/content/en/patterns/media/screen-record/index.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+The example below shows how user's screen can be recorded as webm content, locally previewed on the same page, and saved to the local file system.
```
Comment (tomayac) [inline src/site/content/en/patterns/media/screen-record/index.md:14]:
```suggestion
The example below shows how the user's screen can be recorded in the WebM format, locally previewed on the same page, and saved to the local file system.
```

File: src/site/content/en/patterns/media/screen-record/index.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+```js
+let recorder;
+
+shareScreenButton.addEventListener("click", async () => {
+  // Prompt user to share their screen.
```
Comment (tomayac) [inline src/site/content/en/patterns/media/screen-record/index.md:20]:
```suggestion
  // Prompt the user to share their screen.
```

File: src/site/content/en/patterns/media/screen-record/index.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  const stream = await navigator.mediaDevices.getDisplayMedia();
+  recorder = new MediaRecorder(stream);
+  // Preview screen locally.
```
Comment (tomayac) [inline src/site/content/en/patterns/media/screen-record/index.md:23]:
```suggestion
  // Preview the screen locally.
```

File: src/site/content/en/patterns/media/screen-record/index.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  video.srcObject = stream;
+});
+
+startRecordButton.addEventListener("click", async () => {
+  // For the sake of more legible code, it only uses the
+  // showSaveFilePicker() method. In production, you need to
+  // cater for browsers that don't support this method, as
+  // outlined in https://web.dev/patterns/files/save-a-file/.
```
Comment (tomayac) [inline src/site/content/en/patterns/media/screen-record/index.md:31]:
```suggestion
  // For the sake of more legible code, this sample only uses the
  // `showSaveFilePicker()` method. In production, you need to
  // cater for browsers that don't support this method, as
  // outlined in https://web.dev/patterns/files/save-a-file/.
```

File: src/site/content/en/patterns/media/screen-record/index.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+  // Prompt user to choose where to save the recording file.
```
Comment (tomayac) [inline src/site/content/en/patterns/media/screen-record/index.md:33]:
```suggestion
  // Prompt the user to choose where to save the recording file.
```

File: src/site/content/en/patterns/media/screen-record/index.md [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+  const suggestedName = "screen-record.webm";
```
Comment (tomayac) [inline src/site/content/en/patterns/media/screen-record/index.md:34]:
```suggestion
  const suggestedName = "screen-recording.webm";
```

File: src/site/content/en/patterns/media/screen-record/index.md [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+  const handle = await window.showSaveFilePicker({ suggestedName });
+  const writable = await handle.createWritable();
+
+  // Start recording.
+  recorder.start();
+  recorder.addEventListener("dataavailable", async (event) => {
+    // Write chunks to the file.
+    await writable.write(event.data);
+    if (recorder.state == "inactive") {
```
Comment (tomayac) [inline src/site/content/en/patterns/media/screen-record/index.md:43]:
```suggestion
    if (recorder.state === "inactive") {
```

File: src/site/content/en/patterns/media/screen-record/index.md [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+      // Close file when recording stops.
```
Comment (tomayac) [inline src/site/content/en/patterns/media/screen-record/index.md:44]:
```suggestion
      // Close the file when the recording stops.
```

File: src/site/content/en/patterns/media/screen-record/index.md [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+      await writable.close();
+    }
+  });
+});
+
+stopRecordButton.addEventListener("click", () => {
+  // Stop recording.
```
Comment (tomayac) [inline src/site/content/en/patterns/media/screen-record/index.md:51]:
```suggestion
  // Stop the recording.
```

File: src/site/content/en/patterns/media/screen-record/index.md [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+{% BrowserCompat 'api.Window.showOpenFilePicker' %}
+
+## Further reading
+
+- [W3C Screen Capture Specification](https://w3c.github.io/mediacapture-screen-share/)
+- [W3C MediaStream Recording Specification](https://w3c.github.io/mediacapture-record/#mediarecorder-api)
+- [WICG File System Access Specification](https://wicg.github.io/file-system-access/)
+
+## Demo
```
Comment (tomayac) [inline src/site/content/en/patterns/media/screen-record/index.md:76]:
```suggestion
## Demo

```

File: src/site/content/en/patterns/media/screen-record/index.md [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+
+shareScreenButton.addEventListener("click", async () => {
+  // Prompt user to share their screen.
+  const stream = await navigator.mediaDevices.getDisplayMedia();
+  recorder = new MediaRecorder(stream);
+  // Preview screen locally.
+  video.srcObject = stream;
+});
+
+startRecordButton.addEventListener("click", async () => {
```
Comment (tomayac) [inline src/site/content/en/patterns/media/screen-record/index.md:27]:
```suggestion
stopShareScreenButton.addEventListener("click", () => {
  const tracks = video.srcObject.getTracks();
  tracks.forEach((track) => track.stop());
  video.srcObject = null;
  recorder = null;
});

startRecordButton.addEventListener("click", async () => {
```

File: src/site/content/en/patterns/media/screen-record/assets/body.html [hunk 15]
```diff
@@ -0,0 +1,20 @@
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <meta charset="utf-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1" />
+    <link
+      rel="icon"
+      href="data:image/svg+xml,<svg xmlns=%22http://www.w3.org/2000/svg%22 viewBox=%220 0 100 100%22><text y=%22.9em%22 font-size=%2290%22>🖥️</text></svg>"
+    />
+    <title>How to record screen</title>
```
Comment (tomayac) [inline src/site/content/en/patterns/media/screen-record/assets/body.html:10]:
```suggestion
    <title>How to record the user's screen</title>
```

File: src/site/content/en/patterns/media/screen-record/assets/body.html [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+  </head>
+  <body>
+    <h1>How to record screen</h1>
```
Comment (tomayac) [inline src/site/content/en/patterns/media/screen-record/assets/body.html:13]:
```suggestion
    <h1>How to record the user's screen</h1>
```

File: src/site/content/en/patterns/media/screen-record/assets/body.html [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+    <button id="shareScreenButton">Share screen</button>
```
Comment (tomayac) [inline src/site/content/en/patterns/media/screen-record/assets/body.html:14]:
```suggestion
    <button id="shareScreenButton">Start sharing screen</button>
    <button id="stopShareScreenButton" disabled>Stop sharing screen</button>    
```

File: src/site/content/en/patterns/media/screen-record/assets/script.js [hunk 18]
```diff
@@ -0,0 +1,46 @@
+const video = document.querySelector('video');
+const shareScreenButton = document.querySelector('#shareScreenButton');
```
Comment (tomayac) [inline src/site/content/en/patterns/media/screen-record/assets/script.js:2]:
```suggestion
const shareScreenButton = document.querySelector('#shareScreenButton');
const stopShareScreenButton = document.querySelector('#stopShareScreenButton');
```

File: src/site/content/en/patterns/media/screen-record/assets/script.js [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+let recorder;
+
+shareScreenButton.addEventListener("click", async () => {
+  // Prompt user to share their screen.
+  const stream = await navigator.mediaDevices.getDisplayMedia();
+  recorder = new MediaRecorder(stream);
+  // Preview screen locally.
+  video.srcObject = stream;
+
+  startRecordButton.disabled = false;
```
Comment (tomayac) [inline src/site/content/en/patterns/media/screen-record/assets/script.js:15]:
```suggestion
    stopShareScreenButton.disabled = false;
    startRecordButton.disabled = false;
```

File: src/site/content/en/patterns/media/screen-record/demo.njk [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+      }
+      pre {
+        color: red;
+        white-space: pre-line;
+      }
+    </style>
+  </head>
+  <body>
+    <h1>How to record screen</h1>
+    <button id="shareScreenButton">Share screen</button>
```
Comment (tomayac) [inline src/site/content/en/patterns/media/screen-record/demo.njk:52]:
```suggestion
    <button id="shareScreenButton">Start sharing screen</button>
    <button id="stopShareScreenButton">Stop sharing screen</button>    
```

**Review by tomayac — CHANGES_REQUESTED** [review]
Thanks, just some minor fixes and the suggestion to add a stop screen share button.

(I haven't reflected the proposals throughout, please make the edits everywhere as you work them down. I wish this were less tedious.)

File: src/site/content/en/patterns/media/screen-record/index.md [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+title: How to record the user's screen
+date: 2022-10-14
+authors:
+  - beaufortfrancois
+description: >
+  Learn how to record the user's screen.
+height: 800
+---
+
+Sharing tabs, windows, and screens is possible on the web platform with the [Screen Capture API](https://w3c.github.io/mediacapture-screen-share/). The `getDisplayMedia()` method allows the user to select a screen to capture as a media stream. This stream can then be recorded with the [MediaRecorder API](https://developer.chrome.com/blog/mediarecorder/) or shared with others over the network. The recording can be saved to a local file via the [`showOpenFilePicker()`](/file-system-access/) method.
```
Comment (tomayac) [inline src/site/content/en/patterns/media/screen-record/index.md:12]:
```suggestion
Sharing tabs, windows, and screens is possible on the web platform with the [Screen Capture API](https://w3c.github.io/mediacapture-screen-share/). The [`getDisplayMedia()`](https://developer.mozilla.org/docs/Web/API/MediaDevices/getDisplayMedia) method allows the user to select a screen to capture as a media stream. This stream can then be recorded with the [MediaRecorder API](https://developer.chrome.com/blog/mediarecorder/) or shared with others over the network. The recording can be saved to a local file via the [`showOpenFilePicker()`](https://developer.mozilla.org/docs/Web/API/Window/showOpenFilePicker) method.
```

File: src/site/content/en/patterns/media/screen-record/index.md [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+
+The example below shows how the user's screen can be recorded in the WebM format, locally previewed on the same page, and saved to the local file system.
```
Comment (tomayac) [inline src/site/content/en/patterns/media/screen-record/index.md:14]:
```suggestion
The example below shows how you can record the user's screen in the WebM format, locally preview on the same page, and save the recording to the user's file system.
```

File: src/site/content/en/patterns/media/screen-record/index.md [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+shareScreenButton.addEventListener("click", async () => {
+  // Prompt the user to share their screen.
+  stream = await navigator.mediaDevices.getDisplayMedia();
+  recorder = new MediaRecorder(stream);
+  // Preview the screen locally.
+  video.srcObject = stream;
+});
+
+stopShareScreenButton.addEventListener("click", () => {
+  // Stop stream.
```
Comment (tomayac) [inline src/site/content/en/patterns/media/screen-record/index.md:29]:
```suggestion
  // Stop the stream.
```

**Review by tomayac — APPROVED** [review]
LGTM, approval

Final nit: the "Stop sharing screen" button is not disabled when the user stops sharing. No harm done, but since the other buttons react based on the global state, this might be a dot on an 'i' for the perfectionist ;-)

**Comment by beaufortfrancois** [comment]
@tomayac And we're good!

**Pull request merged**